extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -0,0 +1,735 @@
+        device.chip.remove();
+        device.element.remove();
+      });
+      this.devices_ = this.devices_.filter((d) => d != device);
+      this.updateDevicesInHash_();
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @param {string} deviceName
+   * @return {!Element} the device chip
+   */
+  buildDeviceChip_(deviceName) {
+    const deviceChip = buildDeviceChipTemplate(this.element);
+    deviceChip.querySelector(
+      '.i-amphtml-story-dev-tools-device-chip-name'
+    ).textContent = deviceName;
+    deviceChip.setAttribute('data-action', PREVIEW_ACTIONS.REMOVE_DEVICE);
+    deviceChip.setAttribute('data-device', deviceName);
+    return deviceChip;
+  }
+
+  /**
+   * @param {!Element} chipElement
+   * @private
+   */
+  onDeviceChipToggled_(chipElement) {
+    const deviceName = chipElement.getAttribute('data-device');
+    if (this.removeDevice_(deviceName)) {
+      this.mutateElement(() => chipElement.setAttribute('inactive', ''));
+    } else {
+      this.mutateElement(() => chipElement.removeAttribute('inactive'));
+      this.addDevice_(deviceName);
+    }
+    this.repositionDevices_();
+  }
+
+  /**
+   * If the devices are not the default ones, add them to the hashString
+   * @private
+   */
+  updateDevicesInHash_() {
+    const devicesStringRepresentation = this.devices_
+      .map((device) => simplifyDeviceName(device.name))
+      .join(';');
+    this.element.setAttribute('data-devices', devicesStringRepresentation);
+    updateHash(
+      {
+        'devices':
+          devicesStringRepresentation != DEFAULT_DEVICES
+            ? devicesStringRepresentation
+            : null,
+      },
+      this.win
+    );
+  }
+
+  /**
+   * Recalculate the positions of the devices to keep them spaced and scaled evenly.
+   * @private
+   * */
+  repositionDevices_() {
+    const layoutBox = this.getLayoutBox();
+    let sumDeviceWidths = 0;
+    let maxDeviceHeights = 0;
+    // Find the sum of the device widths and max of heights since they are horizontally laid out.
+    this.devices_.forEach((deviceSpecs) => {
+      sumDeviceWidths += deviceSpecs.width;
+      maxDeviceHeights = Math.max(
+        maxDeviceHeights,
+        deviceSpecs.deviceHeight || deviceSpecs.height + 100
+      );
+    });
+    // Find the scale that covers up to 90% of width or 80% of height.
+    const scale = Math.min(
+      (layoutBox.width / sumDeviceWidths) * 0.9,
+      (layoutBox.height / maxDeviceHeights) * 0.8
+    );
+    const paddingSize =
+      (layoutBox.width - sumDeviceWidths * scale) / (this.devices_.length + 1);
+    let cumWidthSum = paddingSize;
+    this.mutateElement(() => {
+      this.devices_.forEach((deviceSpecs) => {
+        // Calculate the width change when scaling that needs to be added.
+        const scaleWidthChange = deviceSpecs.width * (1 - scale) * 0.5 + 10;
+        setStyles(deviceSpecs.element, {
+          'transform': `perspective(100px) translate3d(${
```

Comment:
Same comment as in `addDevice`:
You could then put `perspective: 100px;` on a wrapper instead of inlining this.

Reply:
Wrote response on the other comment as well, but setting this perspective will have the device scaling center to be the center of the container instead of each device center breaks the current calculations and change the frame of reference in a way that makes it really hard to do the math again.

Reply:
SG. Thanks for explaining this :)